Message flows and endpoint state management (Was: ANNOUNCE) (#643)
moq-wg/moq-transport: draft-ietf-moq-transport.md

expects more OBJECTs to be delivered. The server closes the session with a
'GOAWAY Timeout' if the client doesn't close the session quickly enough.

# Track Discovery and Retrieval (#track-discovery}

> Discovery of MoQT servers is always done out-of-band. Track discovery is done in

Reviewer: s/is done/can be done/ ?

There are cases where track discovery is also out of band (eg catalog, hard coded in specifications), right?

the context of an established MoQT session. The session client might be a

> subscriber, publisher, or both.

Reviewer: > The session client might be a subscriber, publisher, or both.

This is true, but is it related to track discovery?

Given sufficient out of band information, it is valid for a subscriber
to send a SUBSCRIBE or FETCH message to a publisher (including a relay) without

> any previous MoQT messages besides SETUP. However, SUBSCRIBE_NAMESPACES and

Reviewer: SUBSCRIBE_ANNOUNCES

> ANNOUNCE messages provide an in-band means of discovery of subscribers and

Reviewer: I don't think you can discover subscribers via these messages.  

...discovery of publishers and track namespaces?

Aside: I've been thinking there might be value in an optional `track name` (or names) in ANNOUNCE, to assist with catalog-less discovery.

Author: SUBSCRIBE_ANNOUNCES discovers subscribers.

Author: Added "respectively" to make it clearer.

publishers.

The syntax of these messages is described in {{message}}.

## SUBSCRIBE_ANNOUNCES

If the subscriber is aware of a namespace of interest, it can send
SUBSCRIBE_ANNOUNCES to publishers/relays it has discovered. This message

> increases the likelihood that publishers will send relevant ANNOUNCE messages

Reviewer: I don't like 'increases the likelihood'.  Maybe:

The recipient of the message will send any relevant ANNOUNCE messages for that namespace.

for that namespace.

An UNSUBSCRIBE_NAMESPACES negates the effect of a SUBSCRIBE_NAMESPACES. It does
not prohibit the receiver (publisher) from sending further ANNOUNCE messages.
The receiver of a SUBSCRIBE_NAMESPACES_OK or SUBSCRIBE_NAMESPACES_ERROR should
forward the result to the application, so that it can make decisions about

> further publishers to contact. A publisher MUST send exactly one

Reviewer: 'futher publishers or relays' ?

Author: A relay is a publisher.

SUBSCRIBE_NAMESPACES_OK or SUBSCRIBE_NAMESPACES_ERROR. The subscriber SHOULD

> close the session with a protocol error if it detects receiving more than one.

Reviewer: This bit should be moved ahead of the UNSUBSCRIBE_ANNOUNCES description

## ANNOUNCE

A publisher MAY send ANNOUNCE messages to any subscriber. An ANNOUNCE indicates
to the subscriber where to route a SUBSCRIBE or FETCH for that namespace. A
subscriber MAY send SUBSCRIBE or FETCH for a namespace without having received
an ANNOUNCE for it.

If a publisher is authoritative for a given namespace, or is a relay that has

> received an ANNOUNCE for that namespace from an upstream publisher, it MUST send

Reviewer: 'an authorized ANNOUNCE' maybe?

an ANNOUNCE to any subscriber that has subscribed to ANNOUNCE for that

> namespace, a superset of that namespace, or a subset of that namespace. A

Reviewer: Walking through an example:

A: SUBSCRIBE_ANNOUCNES /meeting123
B: SUBSCRIBE_ANNOUNCES /meeting123/video
C: SUBSCRIBE_ANNOUNCES /meeting123/video/afrind

D: ANNOUNCE /meeting123/video

Why does C get ANNOUNCE /meeting123/video?  Given that SUBSCRIBE only matches ANNOUNCE when the namespaces are an *exact* match (not a prefix match), this ANNOUNCE is not relevant to C.

I'm inclined to remove "superset" above.  If the wg changes SUBSCRIBE/ANNOUNCE matching, we can make the matching change here.

> publisher MAY send the ANNOUNCE to any other subscriber.

Reviewer: This MAY opens up the door for "full-mesh" relays that just forward all ANNOUNCEs everywhere.  Maybe we need a little note that indiscriminate broadcasts of ANNOUNCE messages is discouraged?  Or warns of DoS/perf problems?

Author: I think it totally depends on the use case. I personally don't know enough to make a recommendation here.

> However, a publisher SHOULD NOT send an ANNOUNCE advertising a namespace that

Reviewer: nit: Leading with 'However, ' feels jarring, maybe

A publisher SHOULD NOT, however, send ...

exactly matches a namespace for which the subscriber sent an earlier ANNOUNCE
(i.e. an ANNOUNCE ought not to be echoed back to its sender).

> An UNANNOUNCE message negates the meaning of an ANNOUNCE, although it is not a

Reviewer: s/negates the meaning of an/withdraws a previous/

protocol error for the subscriber to send a SUBSCRIBE or FETCH message after
receiving an UNANNOUNCE.

A subscriber can send ANNOUNCE_CANCEL, meaning it is no longer interested in a

> namespace, which also negates the ANNOUNCE: the publisher need not send

Reviewer: I'd say something like:

A subscriber can send ANNOUNCE_CANCEL to indicate cancellation of a previously accepted ANNOUNCE, for example due to expiration of authorization credentials.  The message enables the publisher to ANNOUNCE again with refreshed authorization, or discard associated state.  After receiving an ANNOUNCE_CANCEL, the publisher does not send UNANNOUNCE.

UNANNOUNCE. A publisher and subscriber might seek alternate subscribers and
publishers, respectively, for potential SUBSCRIBE and FETCH interactions.

> The receiver of an ANNOUNCE_OK or ANNOUNCE_ERROR SHOULD report this to the

Reviewer: Should this section move closer to ANNOUNCE (above UNANNOUNCE/ANNOUNCE_CANCEL)

application to inform the search for additional subscribers for a namespace,

> or abandoning the attempt to publish a track. This might be especially useful

Reviewer: > attempt to publish a track

attempt to publish under this namespace

(could be multiple tracks)

in upload or chat applications. A subscriber MUST send exactly one ANNOUNCE_OK
or ANNOUNCE_ERROR in response to an ANNOUNCE. The publisher SHOULD close the
session with a protocol error if it detects receiving more than one.

## SUBSCRIBE/FETCH

A subscriber MAY send a SUBSCRIBE or FETCH for a track to any publisher. If it
has received an ANNOUNCE with a namespace that includes that track, it SHOULD
only request it from the senders of those ANNOUNCE messages.

> The central interaction with a publisher is to send SUBSCRIBE and/or FETCH for

Reviewer: Maybe reorder these to paragraphs?

a particular track in a namespace. The subscriber expects to receive a
SUBSCRIBE_OK/FETCH_OK and objects from the track.

A subscriber keeps SUBSCRIBE state until it sends UNSUBSCRIBE, or after receipt
of a SUBSCRIBE_DONE or SUBSCRIBE_ERROR. Note that SUBSCRIBE_DONE does not
usually indicate that state can immediately be destroyed, see
{{message-subscribe-done}}.

A subscriber keeps FETCH state until it sends FETCH_CANCEL; receives
FETCH_ERROR; or receives a FIN or RESET_STREAM for the FETCH data stream. If the
data stream is already open, it MAY send STOP_SENDING for the data stream along
with FETCH_CANCEL, but MUST send FETCH_CANCEL.

> The Publisher can destroy subscription or fetch state as soon as it has received

Reviewer: The order you have is:

subscriber SUBSCRIBE state
subscriber FETCH state
publisher FETCH state
publisher SUBSCRIBE state

which I find dissatisfying. Can you reorder (either `sub sub/sub fetch/pub sub/pub fetch`, or `sub sub/pub sub/sub fetch/pub fetch` is fine)

Author: That does not accurately describe the publisher paragraphs.

One is via UNSUBSCRIBE/FETCH_CANCEL.

One is SUBSCRIBE_DONE/close the fetch stream.

The first is publisher destroying state to a subscriber message, the second is a local decision.

UNSUBSCRIBE or FETCH_CANCEL, respectively. It MUST reset any open streams
associated with the SUBSCRIBE or FETCH. It can also destroy state after closing
on the FETCH data stream.

The publisher can immediately delete SUBSCRIBE state after sending
SUBSCRIBE_DONE, but MUST NOT send it until it has closed all related streams. It
can destroy all FETCH state after closing the data stream.

A SUBSCRIBE_ERROR or FETCH_ERROR indicates no objects will be delivered, and
both endpoints can immediately destroy relevant state. Objects MUST NOT be sent
for requests that end with an error.

> A publisher MUST send exactly one SUBSCRIBE_OK or SUBSCRIBE_ERROR in response to

Reviewer: Move closer to the top of the section, before the state bits.

expects more OBJECTs to be delivered. The server closes the session with a
'GOAWAY Timeout' if the client doesn't close the session quickly enough.

# Track Discovery and Retrieval (#track-discovery}

Discovery of MoQT servers is always done out-of-band. Track discovery is done in
the context of an established MoQT session. The session client might be a
subscriber, publisher, or both.

Given sufficient out of band information, it is valid for a subscriber
to send a SUBSCRIBE or FETCH message to a publisher (including a relay) without
any previous MoQT messages besides SETUP. However, SUBSCRIBE_NAMESPACES and
ANNOUNCE messages provide an in-band means of discovery of subscribers and
publishers.

The syntax of these messages is described in {{message}}.

## SUBSCRIBE_ANNOUNCES

If the subscriber is aware of a namespace of interest, it can send
SUBSCRIBE_ANNOUNCES to publishers/relays it has discovered. This message

> increases the likelihood that publishers will send relevant ANNOUNCE messages

Author: The normative words are in the ANNOUNCE section. I had SHOULD under the general theory of "server's gonna do..." but I'll change it to MUST because I don't think it matters that much.

for that namespace.

An UNSUBSCRIBE_NAMESPACES negates the effect of a SUBSCRIBE_NAMESPACES. It does

> not prohibit the receiver from sending further ANNOUNCE messages. The receiver

Author: yes.

expects more OBJECTs to be delivered. The server closes the session with a
'GOAWAY Timeout' if the client doesn't close the session quickly enough.

# Track Discovery and Retrieval (#track-discovery}

Discovery of MoQT servers is always done out-of-band. Track discovery is done in
the context of an established MoQT session. The session client might be a
subscriber, publisher, or both.

Given sufficient out of band information, it is valid for a subscriber
to send a SUBSCRIBE or FETCH message to a publisher (including a relay) without
any previous MoQT messages besides SETUP. However, SUBSCRIBE_NAMESPACES and
ANNOUNCE messages provide an in-band means of discovery of subscribers and
publishers.

The syntax of these messages is described in {{message}}.

## SUBSCRIBE_ANNOUNCES

If the subscriber is aware of a namespace of interest, it can send
SUBSCRIBE_ANNOUNCES to publishers/relays it has discovered. This message
increases the likelihood that publishers will send relevant ANNOUNCE messages
for that namespace.

> An UNSUBSCRIBE_NAMESPACES negates the effect of a SUBSCRIBE_NAMESPACES. It does

Reviewer: "negates the effect" has me reaching for a D20

Reviewer: 'withdraws a previous SUBSCRIBE_NAMESPACES' ?

not prohibit the receiver (publisher) from sending further ANNOUNCE messages.
The receiver of a SUBSCRIBE_NAMESPACES_OK or SUBSCRIBE_NAMESPACES_ERROR should
forward the result to the application, so that it can make decisions about
further publishers to contact. A publisher MUST send exactly one
SUBSCRIBE_NAMESPACES_OK or SUBSCRIBE_NAMESPACES_ERROR. The subscriber SHOULD
close the session with a protocol error if it detects receiving more than one.

## ANNOUNCE

A publisher MAY send ANNOUNCE messages to any subscriber. An ANNOUNCE increases

> the likelihood of a subsequent successful SUBSCRIBE or FETCH.

Reviewer: I don't think "increases the likelihood" is the right framing, it's not a +1 modifier to your successful subscribe roll.  ANNOUNCE does what ANNOUNCE does -- it indicates where to route SUBSCRIBE/FETCH that match a namespace.

Reviewer 2: Agree with Alan. Announce acts as anchor point to route subscriptions to as it identifies where the original publisher can be found.

If a publisher is authoritative for a given namespace, or has received an

> ANNOUNCE for that namespace from an upstream publisher, it MUST send an ANNOUNCE

Reviewer: `or *is a relay that* as received`, since generic publishers don't have the concept of upstream publishers.

to any subscriber that has subscribed to ANNOUNCE for that namespace, a

> superset of that namespace, or a subset of that namespace. A publisher MAY send

Reviewer: We should definitely discuss the superset case.

I forget, does a relay route SUBSCRIBE based on exact namespace match pr prefix match?  If it requires an exact match, then what use is knowing who is authoritative for "Foo" if you need to subscribe to ("Foo", "Bar")?

Author: Clearly if I want to know where to get Foo/Bar, the provider of Foo is sufficient?

The text is unclear; I'll add something. Yes, it must be a prefix match to make any sense. all track names are subsets of a namespace; there is no exact match.

the ANNOUNCE to any other subscriber.

However, a publisher SHOULD NOT send an ANNOUNCE advertising a namespace equal

> to, or a subset of, a namespace for which the subscriber sent an earlier ANNOUNCE

Reviewer: I think you mean superset?  But again it depends on exact vs prefix matching.

C->R ANNOUNCE Foo/Bar (I am authoritative for Foo/Bar)
R->C ANNOUNCE Foo (Someone else connected to the relay is authoritative for all of Foo)

(i.e. an ANNOUNCE ought not to be echoed back to its sender).

An UNANNOUNCE message negates the meaning of an ANNOUNCE, although it is not a

> protocol error for the subscriber to send a SUBSCRIBE or FETCH message anyway.

Reviewer: after receiving an UNANNOUNCE

> A subscriber can send ANNOUNCE_CANCEL, meaning it is no longer interested in a

Reviewer: ANNOUNCE_CANCEL is a weird one if it's only function is to prevent UNANNOUNCE?  But if we want to remove it let's not do it in this PR.

Author: It allows the publisher to delete ANNOUNCE state.

namespace, which also negates the ANNOUNCE: the publisher need not send
UNANNOUNCE. A publisher and subscriber might seek alternate subscribers and
publishers, respectively, for potential SUBSCRIBE and FETCH interactions.

The receiver of an ANNOUNCE_OK or ANNOUNCE_ERROR SHOULD report this to the
application to inform the search for additional subscribers for a namespace,
or abandoning the attempt to publish a track. This might be especially useful
in upload or chat applications. A subscriber MUST send exactly one ANNOUNCE_OK
or ANNOUNCE_ERROR in response to an ANNOUNCE. The publisher SHOULD close the

> session with a protocol error if it detects receiving more than one.

Reviewer: Why not MUST?

Author: It might not detect it. When I get ANNOUNCE_ERROR the state is gone. If another ANNOUNCE_ERROR or _OK arrives, I don't know if it's a protocol error or a response to an ANNOUNCE where I already sent UNANNOUNCE and nuked the state.

I don't believe in forcing endpoints to keep a bunch of state just to enforce protocol errors, unless they are security risks.

## SUBSCRIBE/FETCH

A subscriber MAY send a SUBSCRIBE or FETCH for a track to any publisher. If it
has received an ANNOUNCE with a namespace that includes that track, it SHOULD
only request it from the senders of those ANNOUNCE messages.

The central interaction with a publisher is to send SUBSCRIBE and/or FETCH for
a particular track in a namespace. The subscriber expects to receive a
SUBSCRIBE_OK/FETCH_OK and objects from the track.

> A subscriber keeps SUBSCRIBE state until it sends UNSUBSCRIBE, or after receipt

Reviewer: Do we need to mention SUBSCRIBE_ERROR?

of a SUBSCRIBE_DONE.  Note that SUBSCRIBE_DONE does not usually indicate that
state can immediately be destroyed, see {{message-subscribe-done}}.

> A subscriber keeps FETCH state until it sends FETCH_CANCEL, or a FIN or

Reviewer: Do we need to mention FETCH_ERROR?

RESET_STREAM for the FETCH data stream. If the data stream is already open, it

> MAY send STOP_SENDING for the data stream, but MUST send FETCH_CANCEL.

Reviewer: Why does it have to send FETCH_CANCEL?

SHOULD send STOP_SENDING?

Author: As you mentioned in a different issue, in case the data stream isn't open yet. The state of the stream may not be the same at both endpoints, so this makes it easier.

FETCH_CANCEL results in RESET_STREAM, just the same as STOP_SENDING. so the advantage is... what? that the QUIC control message is slightly faster than an MoQT control message?

The Publisher can destroy subscription or fetch state as soon as it has received

> UNSUBSCRIBE or FETCH_CANCEL, respectively. It MUST reset any open streams

Reviewer: Why not STOP_SENDING on the fetch stream?

Reviewer: Do we talk about STOP_SENDING on subscribe streams anywhere?   Seems like those should elicit resets but shouldn't affect subscribe state.

Author: I added that it can delete state after closing the data stream, which is a more general case than getting STOP_SENDING.

STOP_SENDING is irrelevant to SUBSCRIBE state. If it's not discussed, file a new issue.

associated with the SUBSCRIBE or FETCH.

The publisher can immediately delete SUBSCRIBE state after sending
SUBSCRIBE_DONE, but MUST NOT send it until it has closed all related streams. It
can destroy all FETCH state after closing the data stream.

A SUBSCRIBE_ERROR or FETCH_ERROR indicates no objects will be delivered, and
both endpoints can immediately destroy relevant state. Objects MUST NOT be sent
for requests that end with an error.

A publisher MUST send exactly one SUBSCRIBE_OK or SUBSCRIBE_ERROR in response to
a SUBSCRIBE. It MUST send exactly one FETCH_OK or FETCH_ERROR in response to a

> FETCH. The subscriber SHOULD close the session with a protocol error if it

Reviewer: MUST

Author: Again, my implementation will not detect a second message after _ERROR.

expects more OBJECTs to be delivered. The server closes the session with a
'GOAWAY Timeout' if the client doesn't close the session quickly enough.

# Track Discovery and Retrieval (#track-discovery}

Discovery of MoQT servers is always done out-of-band. Namespace discovery can be
done in the context of an established MoQT session.

Given sufficient out of band information, it is valid for a subscriber
to send a SUBSCRIBE or FETCH message to a publisher (including a relay) without
any previous MoQT messages besides SETUP. However, SUBSCRIBE_ANNOUNCES and

> ANNOUNCE messages provide an in-band means of discovery of subscribers and

Reviewer 2: I still not sure if we can describe subscriber though ;)

Author: I explicity defined the term.

Reviewer 2: not sure how can discovery subscribers using announces/subscribe_announces.  These messages allow a way to discover a publisher(s) for a given namespace/namespace-prefix and they aid in routing a subscription to the publishers.

Author: An endpoint that receives a SUBSCRIBE_ANNOUNCES can now identify the sender as a subscriber.

publishers, respectively.

The syntax of these messages is described in {{message}}.

## SUBSCRIBE_ANNOUNCES

If the subscriber is aware of a namespace of interest, it can send
SUBSCRIBE_ANNOUNCES to publishers/relays it has discovered. The recipient of
this message will send any relevant ANNOUNCE messages for that namespace, or
subset of that namespace.

A publisher MUST send exactly one SUBSCRIBE_NAMESPACES_OK or
SUBSCRIBE_NAMESPACES_ERROR. The subscriber SHOULD close the session with a
protocol error if it detects receiving more than one.

An UNSUBSCRIBE_NAMESPACES withdraws a previous SUBSCRIBE_NAMESPACES. It does
not prohibit the receiver (publisher) from sending further ANNOUNCE messages.

> The receiver of a SUBSCRIBE_NAMESPACES_OK or SUBSCRIBE_NAMESPACES_ERROR should

Author: I changed it to "ought to".

I'm surprised to hear you think it would be handled internally. Within a session, the decision to subscribe to a namespace has to come from the application, no? And if rejected, the application is going to have to initiate an MoQT session with someone else (or fix their authentication)?

> The receiver of a SUBSCRIBE_NAMESPACES_OK or SUBSCRIBE_NAMESPACES_ERROR should

Reviewer 3: I guess, but you can also SUBSCRIBE without ever receiving an ANNOUNCE, so I'm having a difficult time reasoning about how ANNOUNCE will be used.  But I'm fine with keeping it as is for now.

> The receiver of a SUBSCRIBE_NAMESPACES_OK or SUBSCRIBE_NAMESPACES_ERROR should

Author: Sure, but if the app sends a SUBSCRIBE_ANNOUNCE surely it needs one, and needs to know if the one it asked for failed?

SUBSCRIBE_OK/FETCH_OK and objects from the track.

A subscriber MAY send a SUBSCRIBE or FETCH for a track to any publisher. If it
has received an ANNOUNCE with a namespace that exactly matches the namespace for
that track, it SHOULD only request it from the senders of those ANNOUNCE
messages.

A publisher MUST send exactly one SUBSCRIBE_OK or SUBSCRIBE_ERROR in response to
a SUBSCRIBE. It MUST send exactly one FETCH_OK or FETCH_ERROR in response to a
FETCH. The subscriber SHOULD close the session with a protocol error if it
detects receiving more than one.

A subscriber keeps SUBSCRIBE state until it sends UNSUBSCRIBE, or after receipt
of a SUBSCRIBE_DONE or SUBSCRIBE_ERROR. Note that SUBSCRIBE_DONE does not
usually indicate that state can immediately be destroyed, see
{{message-subscribe-done}}.

A subscriber keeps FETCH state until it sends FETCH_CANCEL; receives
FETCH_ERROR; or receives a FIN or RESET_STREAM for the FETCH data stream. If the
data stream is already open, it MAY send STOP_SENDING for the data stream along
with FETCH_CANCEL, but MUST send FETCH_CANCEL.

The Publisher can destroy subscription or fetch state as soon as it has received

> UNSUBSCRIBE or FETCH_CANCEL, respectively. It MUST reset any open streams

Reviewer: > It MUST reset

Is this new normative language and if so it doesn't seem very ANNOUNCE related.  That said it seems like the right behavior -- if you want less abrupt termination, use SUBSCRIBE_UPDATE?

Author: This whole section is about managing state related to SUBSCRIBE_ANNOUNCES, ANNOUNCE, SUBSCRIBE and FETCH. I don't see how this bit can be sensibly separated.

Reviewer: Maybe retitle PR to "Message flows and endpoint state management"?

associated with the SUBSCRIBE or FETCH. It can also destroy state after closing

> on the FETCH data stream.

Reviewer 3: ```suggestion
the FETCH data stream.
```

expects more OBJECTs to be delivered. The server closes the session with a
'GOAWAY Timeout' if the client doesn't close the session quickly enough.

# Track Discovery and Retrieval (#track-discovery}

Discovery of MoQT servers is always done out-of-band. Track discovery can be

> done in the context of an established MoQT session.

Reviewer 3: Maybe I'm confused, but I think you can discover announced namespaces, but not tracks, correct?

Given sufficient out of band information, it is valid for a subscriber
to send a SUBSCRIBE or FETCH message to a publisher (including a relay) without
any previous MoQT messages besides SETUP. However, SUBSCRIBE_ANNOUNCES and
ANNOUNCE messages provide an in-band means of discovery of subscribers and
publishers, respectively.

The syntax of these messages is described in {{message}}.

## SUBSCRIBE_ANNOUNCES

If the subscriber is aware of a namespace of interest, it can send
SUBSCRIBE_ANNOUNCES to publishers/relays it has discovered. The recipient of
this message will send any relevant ANNOUNCE messages for that namespace, or
subset of that namespace.

A publisher MUST send exactly one SUBSCRIBE_ANNOUNCES_OK or
SUBSCRIBE_ANNOUNCES_ERROR  in response to a SUBSCRIBE_ANNOUNCES.

> The subscriber SHOULD close the session with a protocol error if it detects receiving
> more than one.

Reviewer 3: ```suggestion
The subscriber SHOULD close the session with a protocol error if it detects
receiving more than one.
```

to the subscriber where to route a SUBSCRIBE or FETCH for that namespace. A
subscriber MAY send SUBSCRIBE or FETCH for a namespace without having received
an ANNOUNCE for it.

If a publisher is authoritative for a given namespace, or is a relay that has
received an authorized ANNOUNCE for that namespace from an upstream publisher,
it MUST send an ANNOUNCE to any subscriber that has subscribed to ANNOUNCE for
that namespace, or a subset of that namespace. A publisher MAY send the ANNOUNCE
to any other subscriber.

A publisher SHOULD NOT, however, send an ANNOUNCE advertising a namespace that
exactly matches a namespace for which the subscriber sent an earlier ANNOUNCE
(i.e. an ANNOUNCE ought not to be echoed back to its sender).

The receiver of an ANNOUNCE_OK or ANNOUNCE_ERROR SHOULD report this to the
application to inform the search for additional subscribers for a namespace,
or abandoning the attempt to publish under this namespace. This might be
especially useful in upload or chat applications. A subscriber MUST send exactly
one ANNOUNCE_OK or ANNOUNCE_ERROR in response to an ANNOUNCE. The publisher

> SHOULD close the session with a protocol error if it detects receiving more than
> one.

Reviewer 3: ```suggestion
SHOULD close the session with a protocol error if it receives more than one.
```

## Subscriber Interactions

Subscribers interact with the Relays by sending a SUBSCRIBE or FETCH
control message for the tracks of interest. Relays MUST ensure subscribers are
authorized to access the content associated with the track. The authorization
information can be part of subscription request itself or part of the
encompassing session. The specifics of how a relay authorizes a user are outside
the scope of this specification.

The relay will have to send an upstream SUBSCRIBE and/or FETCH if it does not
have all the objects in the FETCH, or is not currently subscribed to the full
requested range. In this case, it MUST withhold sending its own SUBSCRIBE_OK or

> FETCH_OK until receiving one from upstream.

Reviewer: Let's make this change in a separate PR, since it's not ANNOUNCE related.

Author: This is eliminating duplicate text, with some cleanup at the same time.

Author: Moved to #645

## Subscriber Interactions

> Subscribers interact with the Relays by sending a SUBSCRIBE
> ({{message-subscribe-req}}) control message for the tracks of
> interest. Relays MUST ensure subscribers are authorized to access the

Reviewer 3: ```suggestion
Subscribers subscribe to tracks by sending  a SUBSCRIBE
({{message-subscribe-req}}) control message for each track of
interest. Relays MUST ensure subscribers are authorized to access the
```

If multiple objects are received with the same Full Track Name,
Group ID and Object ID, Relays MAY ignore subsequently received Objects
or MAY use them to update the cache. Implementations that update the
cache need to protect against cache poisoning.

> Caching can reduce the number of upstream FETCH requests.

Reviewer: Where did this come from -- also not ANNOUNCE

Author: It seems silly to have a separate PR for this editorial line.

Author: moved to #645

A relay MUST NOT reorder or drop objects received on a multi-object stream when
forwarding to subscribers, unless it has application specific information.

Relays SHOULD aggregate authorized subscriptions for a given track when
multiple subscribers request the same track, using SUBSCRIBE_UPDATE when new

> subscriptions change the aggregate properties. Subscription aggregation

Reviewer: Not ANNOUNCE

Author: this is just housekeeping. There is already a ton of editing in this section.

Author: reverted, because I don't think SHOULD/MAY matters.

or MAY use them to update the cache. Implementations that update the
cache need to protect against cache poisoning.

A relay MUST NOT reorder or drop objects received on a multi-object stream when
forwarding to subscribers, unless it has application specific information.

Relays SHOULD aggregate authorized subscriptions for a given track when

> multiple subscribers request the same track, using SUBSCRIBE_UPDATE when new

Reviewer 3: Is the "using SUBSCRIBE_UPDATE..." clause useful without explanation about how/when to use it?

Author: Well, it says "when new subscriptions change the aggregate properties"

I was going to use the example of a new subscription having a  later end group, and therefore sending SUBSCRIBE_UPDATE. But of course that's illegal! So I just eliminated the clause.

ANNOUNCE_CANCEL.

A relay manages sessions from multiple publishers and subscribers,

> connecting them based on the track namespace. This MUST use an exact

Reviewer: You removed the normative exact match

Author: Multi-tiered namespaces don't work if we don't update this.

Author: I thought this was uncontroversial, but I guess not. I'll just revert it, because I don't really care

Publishing through the relay starts with publisher sending ANNOUNCE
control message with a `Track Namespace` ({{model-track}}).
The ANNOUNCE enables the relay to know which publisher to forward a
SUBSCRIBE to.

> Relays MUST ensure that publishers are authorized by verifying that the
> publisher is authorized to publish the content associated with the set of

Reviewer 3: ```suggestion
Relays MUST verify that publishers are authorized to publish
the content associated with the set of
```

ANNOUNCE_CANCEL.

A relay manages sessions from multiple publishers and subscribers,
connecting them based on the track namespace. This MUST use an exact
match on track namespace unless otherwise negotiated by the application.
For example, a SUBSCRIBE namespace=foobar message will be forwarded to

> the session that sent ANNOUNCE namespace=foobar.

Reviewer: Did we lose the example?

A subscriber issues a `UNSUBSCRIBE` message to a publisher indicating it is no
longer interested in receiving media for the specified track and Objects

> should stop being sent as soon as possible.

Reviewer: non-normative should

Reviewer: Still there

* Reason Phrase: Provides the reason for announcement error.

The application SHOULD use a relevant error code in an ANNOUNCE_ERROR, as

> defined below.

Reviewer: Maybe just "the following error codes are defined"?

* Reason Phrase: Provides the reason for announcement error.

> The following error codes are defined:

Reviewer: I'm not going to die on this hill, but moving the error code tables could easily go in another PR.

  the subscriber MUST close the connection with a Duplicate Track Alias error
  ({{session-termination}}).

> The application SHOULD use a relevant error code in SUBSCRIBE_ERROR, as defined below:

Reviewer: SHOULD seems silly here

* Reason Phrase: Provides the reason for fetch error.

> The application SHOULD use a relevant error code in FETCH_ERROR, as defined below:

Reviewer: And here
